pkg/pool-weighted/contracts/InvariantGrowthProtocolFees.sol

```diff
@@ -0,0 +1,81 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "./BaseWeightedPool.sol";
+
+abstract contract InvariantGrowthProtocolFees is BaseWeightedPool {
+    using FixedPoint for uint256;
+
+    // This Pool pays protocol fees by measuring the growth of the invariant between joins and exits. Since weights are
+    // immmutable, the invariant only changes due to accumulated swap fees, which lets the Pool not perform any protocol
+    // fee associated computation or accounting during swaps, saving gas.
+    // This mechanism requires keeping track of the invariant after the last join or exit.
+    uint256 private _lastPostJoinExitInvariant;
+
+    /**
+     * @dev Returns the value of the invariant after the last join or exit operation.
+     */
+    function getLastInvariant() public view returns (uint256) {
+        return _lastPostJoinExitInvariant;
+    }
+
+    function _beforeJoinExit(
+        uint256[] memory preBalances,
+        uint256[] memory normalizedWeights,
+        uint256 protocolSwapFeePercentage
+    ) internal virtual override {
+        // Before joins and exits, we measure the growth of the invariant compared to the invariant after the last join
+        // or exit, which will have been caused by swap fees, and use it to mint BPT as protocol fees. This dillutes all
+        // LPs, which means that new LPs will join the pool debt-free, and LPs exiting will pay their due befere
+        // leaving.
+
+        // We return immediately if the fee percentage is zero (to avoid unnecessary computation), or when the pool is
+        // paused (to avoid complex computation during emergency withdrawals).
+        if ((protocolSwapFeePercentage == 0) || !_isNotPaused()) {
+            return;
+        }
+
+        uint256 preJoinExitInvariant = WeightedMath._calculateInvariant(normalizedWeights, preBalances);
+
+        uint256 toMint = WeightedMath._calcDueProtocolSwapFeeBPTAmount(
+            totalSupply(),
+            _lastPostJoinExitInvariant,
+            preJoinExitInvariant,
+            protocolSwapFeePercentage
+        );
+
+        _payProtocolFees(toMint);
+    }
+
+    function _afterJoinExit(
+        bool isJoin,
+        uint256[] memory preBalances,
+        uint256[] memory balanceDeltas,
+        uint256[] memory normalizedWeights
+    ) internal virtual override {
+        // After all joins and exits we store the post join/exit invariant in order to compute growth due to swap fees
+        // in the next one.
+
+        // Compute the post balances by adding or removing the deltas. Note that we're allowed to mutate preBalances.
+        for (uint256 i = 0; i < preBalances.length; ++i) {
+            preBalances[i] = isJoin ? preBalances[i].add(balanceDeltas[i]) : preBalances[i].sub(balanceDeltas[i]);
```

Comment:
😛 

```suggestion
            preBalances[i] = (isJoin ? SafeMath.add : SafeMath.sub)(preBalances[i], balanceDeltas[i]);
```

Reply:
Oh wow I did not know you could do this!